libvirt/domain.go

```diff
@@ -706,23 +706,59 @@ func setNetworkInterfaces(d *schema.ResourceData, domainDef *libvirtxml.Domain,
 		}
 
 		// connect to the interface to the network... first, look for the network
-		if n, ok := d.GetOk(prefix + ".network_name"); ok {
-			// when using a "network_name" we do not try to do anything: we just
-			// connect to that network
-			netIface.Source = &libvirtxml.DomainInterfaceSource{
-				Network: &libvirtxml.DomainInterfaceSourceNetwork{
-					Network: n.(string),
-				},
+		var network *libvirt.Network
+
+		if networkName, ok := d.GetOk(prefix + ".network_name"); ok {
+			n, err := virConn.LookupNetworkByName(networkName.(string))
+			if err != nil {
+				return fmt.Errorf("Can't retrieve network '%s'", networkName.(string))
 			}
+			network = n
+			defer network.Free()
+
 		} else if networkUUID, ok := d.GetOk(prefix + ".network_id"); ok {
 			// when using a "network_id" we are referring to a "network resource"
 			// we have defined somewhere else...
-			network, err := virConn.LookupNetworkByUUIDString(networkUUID.(string))
+			n, err := virConn.LookupNetworkByUUIDString(networkUUID.(string))
 			if err != nil {
 				return fmt.Errorf("Can't retrieve network ID %s", networkUUID)
 			}
+			network = n
 			defer network.Free()
 
+		} else if bridgeNameI, ok := d.GetOk(prefix + ".bridge"); ok {
+			netIface.Source = &libvirtxml.DomainInterfaceSource{
+				Bridge: &libvirtxml.DomainInterfaceSourceBridge{
+					Bridge: bridgeNameI.(string),
+				},
+			}
+		} else if devI, ok := d.GetOk(prefix + ".vepa"); ok {
+			netIface.Source = &libvirtxml.DomainInterfaceSource{
+				Direct: &libvirtxml.DomainInterfaceSourceDirect{
+					Dev:  devI.(string),
+					Mode: "vepa",
+				},
+			}
+		} else if devI, ok := d.GetOk(prefix + ".macvtap"); ok {
+			netIface.Source = &libvirtxml.DomainInterfaceSource{
+				Direct: &libvirtxml.DomainInterfaceSourceDirect{
+					Dev:  devI.(string),
+					Mode: "bridge",
+				},
+			}
+		} else if devI, ok := d.GetOk(prefix + ".passthrough"); ok {
+			netIface.Source = &libvirtxml.DomainInterfaceSource{
+				Direct: &libvirtxml.DomainInterfaceSourceDirect{
+					Dev:  devI.(string),
+					Mode: "passthrough",
+				},
+			}
+		} else {
+			// no network has been specified: we are on our own
+		}
+
+		// if we got a network
+		if network != nil {
 			networkName, err := network.GetName()
 			if err != nil {
 				return fmt.Errorf("Error retrieving network name: %s", err)
```

Comment:
Before the `return` statement:

```golang
defer network.Free()
```

The `network` instance has not been added to the `partialNetIFace` map, hence we can safely release the object right now. There's no risk of it being free-ed up again as a mistake.

Maybe I would add these lines as comments inside of the block.

Reply:
I don't see the advantage of doing this. The idea of using the `Ref()` call is to increment the reference counter only if the network is added to partialNetIFace, is that in this way, it is not necessary to consider carefully all the cases when it can be freed. It can always be freed. I think the logic is more straightforward in this way.

Reply:
Additionally, if there is more than one interface on the same network that must be added to `partialNetIface`, it is much easier to keep track of the usage of the network using `Ref`to increment the reference count and `Free` it independently for each interface.